calicoctl/commands/checksystem.go

```diff
@@ -0,0 +1,257 @@
+// Copyright (c) 2016 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package commands
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+
+	docopt "github.com/docopt/docopt-go"
+	goversion "github.com/mcuadros/go-version"
+)
+
+// The minimum allowed linux kernel version is 2.6.24, which introduced network
+// namespaces and veth pairs.
+const minKernelVersion = "2.6.24"
+
+// Required kernel modules to run Calico
+var requiredModules = []string{"xt_set", "ip6_tables"}
+
+// Checksystem checks host system for compatible versions
+func Checksystem(args []string) error {
+	doc := `Usage: 
+    calicoctl checksystem
+
+Description:
+    Check for incompatibilities between Calico and the host system`
+
+	parsedArgs, err := docopt.Parse(doc, args, true, "calicoctl", false, false)
+	if err != nil {
+		return err
+	}
+
+	// Return if no args are passed or `-h` `--help` is passed
+	if len(parsedArgs) == 0 {
+		return nil
+	}
+
+	// Make sure the command is run with super user priviladges
+	if os.Getuid() != 0 {
+		fmt.Println("Need super user privilages: Operation not permitted")
+		os.Exit(1)
+	}
+
+	systemOk := true
+
+	fmt.Print("Checking kernel version...\n")
+	err = checkKernelVersion()
+	if err != nil {
+		systemOk = false
+	}
+
+	fmt.Print("Checking kernel modules...\n")
+	err = checkKernelModules()
+	if err != nil {
+		systemOk = false
+	}
+
+	// If any of the checks fail, print a message and exit
+	if !systemOk {
+		fmt.Printf("%-20s\t\t\t%s", "System doesn't meet one or more minimum systems requirements to run Calico", "\u2717\n")
+		os.Exit(1)
+	}
+
+	fmt.Printf("%-20s\t\t\t%s", "System meets minimum system requirements to run Calico!", "\u2713\n")
+
+	return nil
+}
+
+// checkKernelVersion checks for minimum required kernel version
+func checkKernelVersion() error {
+	kernelVersion, err := exec.Command("uname", "-r").Output()
+	if err != nil {
+		log.Println(err)
+	}
+
+	// To strip the trailing `\n`
+	kernelVersionStr := strings.TrimSpace(string(kernelVersion))
+
+	// Print the version for the command output
+	fmt.Printf("\t\t%-20s", kernelVersionStr)
+
+	// Goversion normalizes the versions and compares them returns `true` if
+	// running version is >= minimum version
+	if !goversion.CompareNormalized(kernelVersionStr, minKernelVersion, ">=") {
+
+		// Prints "✗" if current version is not >= minimum required version
+		fmt.Printf("\t\t\t\t\t\u2717\n")
```

Comment:
you're right, didn't think about telnet. I'll replace ✓ with `OK.` and ✗ with `FAIL`